src/analysis/processing/qgsalgorithmrandompointsinpolygons.cpp

```diff
@@ -0,0 +1,399 @@
+      emptyOrNullGeom++;
+      featureCount++;
+      baseFeatureProgress += featureProgressStep;
+      feedback->setProgress( baseFeatureProgress );
+      continue;
+    }
+    QgsGeometry polyGeom( polyFeat.geometry() );
+    if ( polyGeom.isEmpty() )
+    {
+      // Increment invalid features count
+      emptyOrNullGeom++;
+      featureCount++;
+      baseFeatureProgress += featureProgressStep;
+      feedback->setProgress( baseFeatureProgress );
+      continue;
+    }
+    if ( mDynamicNumPoints || mDynamicMinDistance || mDynamicMaxAttempts )
+    {
+      expressionContext.setFeature( polyFeat );
+    }
+    // Initialize the local (per polygon) index
+    ::localIndex2 = QgsSpatialIndex();
+    ::mLocalIndexPoints = 0;
+    // double lineLength = polyGeom.length();
+    int pointsAddedForThisFeature = 0;
+    int numberPointsForThisFeature = mNumPoints;
+    if ( mDynamicNumPoints )
+      numberPointsForThisFeature = mNumPointsProperty.valueAsInt( expressionContext, numberPointsForThisFeature );
+    desiredNumberOfPoints += numberPointsForThisFeature;
+    int maxAttemptsForThisFeature = mMaxAttempts;
+    if ( mDynamicMaxAttempts )
+      maxAttemptsForThisFeature = mMaxAttemptsProperty.valueAsInt( expressionContext, maxAttemptsForThisFeature );
+    ::minDistanceForThisFeature = mMinDistance;
+    if ( mDynamicMinDistance )
+      ::minDistanceForThisFeature = mMinDistanceProperty.valueAsDouble( expressionContext, ::minDistanceForThisFeature );
+    const double pointProgressIncrement = featureProgressStep / ( numberPointsForThisFeature * maxAttemptsForThisFeature );
+    double pointProgress = 0.0;
+    if ( ( ::minDistanceForThisFeature == 0 ) && ( ::mMinDistanceGlobal == 0 ) )
+    {
+      unsigned long fseed = uniformIntDist( mt );
+      QVector< QgsPointXY > newPoints = polyGeom.randomPointsInPolygon( numberPointsForThisFeature, fseed );
+      for ( int i = 0; i < newPoints.length(); i++ )
+      {
+        // add the point
+        QgsPointXY pt = newPoints[i];
+        QgsFeature f = QgsFeature( totNPoints );
+        QgsAttributes pAttrs = QgsAttributes();
+        pAttrs.append( totNPoints );
+        if ( mIncludePolygonAttr )
+        {
+          pAttrs.append( polyFeat.attributes() );
+        }
+        f.setAttributes( pAttrs );
+        QgsGeometry newGeom = QgsGeometry::fromPointXY( pt );
+        f.setGeometry( newGeom );
+        sink->addFeature( f, QgsFeatureSink::FastInsert );
+        totNPoints++;
+        pointsAddedForThisFeature++;
+        pointProgress += pointProgressIncrement * ( maxAttemptsForThisFeature );
+      }
+      feedback->setProgress( baseFeatureProgress + pointProgress );
+      continue;
+    }
+    else
+    {
+      // Have to check for minimum distance
+      unsigned long fseed = uniformIntDist( mt );
+      QVector< QgsPointXY > newPoints = polyGeom.randomPointsInPolygon( numberPointsForThisFeature, pointOK, fseed, feedback, maxAttemptsForThisFeature );
```

Comment:
What we should do here is use a lambda function and remove all the global variables/ methods. 

E.g.

```suggestion
      QVector< QgsPointXY > newPoints = polyGeom.randomPointsInPolygon( numberPointsForThisFeature, [=]( const QgsPointXY &newPoint ) -> bool {
       
       .... contents of existing pointOK function sit here      
      
      }, fseed, feedback, maxAttemptsForThisFeature );
```